fn/mvar.go

```diff
@@ -0,0 +1,205 @@
+//
+// NewMVar : A -> MVar[A].
+func NewMVar[A any](a A) MVar[A] {
+	z := Zero[A]()
+	z.value.Store(&a)
+	return z
+}
+
+// Take will wait for a value to be put into the MVar and then immediately
+// take it out.
+//
+// Take : MVar[A] -> A.
+func (m *MVar[A]) Take() A {
+	var val *A
+
+	// Here we use a Cond pattern and try to Swap nil into value. If we get
+	// a non-nil out it means we can actually finish the Take.
+	m.mutex.Lock()
+	for val = m.value.Swap(nil); val == nil; val = m.value.Swap(nil) {
+		m.tCond.Wait()
+	}
+	m.mutex.Unlock()
+
+	// now that we've successfully swapped a non-nil value out of the
+	// atomic, we can now wake any Put operations which would replace the
+	// value we just removed.
+	m.pCond.Signal()
+
+	return *val
+}
+
+// TryTake is the non-blocking version of TakeMVar, it will return an
+// None() Option if it would have blocked.
+//
+// TryTake : MVar[A] -> Option[A].
+func (m *MVar[A]) TryTake() Option[A] {
+	// If Swap is non-nil it means we can actually Take a value out.
+	if val := m.value.Swap(nil); val != nil {
+		// We just managed to Swap a nil in so we need to give blocked
+		// Puts the opportunity to proceed.
+		m.pCond.Signal()
+
+		return Some(*val)
+	}
+
+	return None[A]()
+}
+
+// Put will wait for a value to be made empty and will immediately replace it
+// with the argument.
+//
+// Put : (MVar[A], A) -> ().
+func (m *MVar[A]) Put(a A) {
+	// Here we use a Cond pattern and try to CompareAndSwap expecting value
+	// to be nil. If we successfully swap, it means we can actually finish
+	// the Put.
+	m.mutex.Lock()
+	swapped := m.value.CompareAndSwap(nil, &a)
+	for ; !swapped; swapped = m.value.CompareAndSwap(nil, &a) {
+		m.pCond.Wait()
+	}
+	m.mutex.Unlock()
+
+	// Now that we set value, we wake the Reads.
+	m.rCond.Broadcast()
+
+	// Here we wait for all the Reads to finish.
+	m.reads.Wait()
+
+	// Now that the readers have finished we wake a single Take.
+	m.tCond.Signal()
+}
+
+// TryPut is the non-blocking version of Put and will return true if the MVar is
+// successfully set.
+//
+// TryPut : (MVar[A], A) -> bool.
+func (m *MVar[A]) TryPut(a A) bool {
+	// We use CompareAndSwap expecting value to be nil indicating an empty
+	// state.
+	if !m.value.CompareAndSwap(nil, &a) {
+		return false
+	}
+
+	// Now that we set value, we wake the Reads.
+	m.rCond.Broadcast()
+
+	// Here we wait for all the Reads to finish.
+	m.reads.Wait()
+
+	// Now that the readers have finished we wake a single Take.
+	m.tCond.Signal()
+
+	return true
+}
+
+// Read will atomically read the contents of the MVar. If the MVar is empty,
+// Read will block until a value is put in. Callers of Read are guaranteed to
+// be woken up before callers of Take.
+//
+// Read : MVar[A] -> A.
+func (m *MVar[A]) Read() A {
+	var val *A
+
+	// While the Read is going we need to increment the WaitGroup so that
+	// a waiting Put operation knows when it can actually wake any waiting
+	// Takes
+	m.reads.Add(1)
```

Comment:
Still getting through the diff, but: is this wait group ever reset? We'll need to take care to ensure that the wait group can't have a `Add` call follow a `Wait`, otherwise we'll run into a panic.

Reply:
The only call to wait is in Put, and it is just to ensure that all blocked readers are satisfied before control is handed to a blocked Take operation.

Reply:
Turns out that an interleaving stress test revealed this was an issue. It has been fixed now